### PR TITLE
Cookie hardening

### PR DESCRIPTION
**Pull request by jonasfj**

### CODE REVIEW

File: app/lib/frontend/handlers/account.dart [hunk 1]
```diff
@@ -42,12 +42,25 @@ Future<shelf.Response> updateSessionHandler(shelf.Request request,
...
+      // Hence, such a cookie cannot have been set by another website or an
+      // HTTP proxy for this website.
+      '$pubSessionCookieName=${newSession.sessionId}',
+      // Send cookie to anything under '/' required by '__Host-' prefix.
+      'Path=/',
+      // Do not include the cookie in CORS requests, unless the request is a
+      // top-level navigation to the site, as recommended in:
+      // https://tools.ietf.org/html/draft-ietf-httpbis-rfc6265bis-02#section-8.8.2
+      'SameSite=Lax',
+      'Secure', // Only allow this cookie to be sent when making HTTPS requests.
```
Comment (isoos) [inline app/lib/frontend/handlers/account.dart]: can we have a conditional for localhost?
e.g. 
```
if (request.requestedUri.scheme != 'http') 'Secure',
```
(or even `if (!request.requestedUri.toString().startsWith('http://localhost:8080/') 'Secure',`

Comment (jonasfj) [reply app/lib/frontend/handlers/account.dart]: I think we should somehow put into the `configuration` whether the process was started with `--local-dev` and only in that case should we do something special.

People might be hitting the service with `Host: localhost` in production, you never know. Certainly there is no good reason to allow things like that in production. Let's be defensive :)

Comment (isoos) [reply app/lib/frontend/handlers/account.dart]: ++

**Comment by jonasfj** [comment]
@isoos, please take a look, I did a lot of changes since last review.. Now it should work on staging and on localhost (provided you're using the right host).

I also added more hosts to the oauth2 flow for our staging site, so we can use the version specific host name.

**Comment by jonasfj** [comment]
This currently deployed to `https://jonasfj-dot-dartlang-pub-dev.appspot.com`

File: app/lib/account/session_cookie.dart [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+/// Parse [cookieString] and return `sessionId` or `null`.
+///
+/// The [cookieString] is the value of the `cookie:` request header.
+String parseSessionCookie(String cookieString) {
+  ArgumentError.checkNotNull(cookieString, 'cookieString');
+
+  // The cookieString is separated by '; ', and contains 'name=value'
+  // See: https://tools.ietf.org/html/rfc6265#section-5.4
+  final cookie = cookieString.split('; ').firstWhere(
```
Comment (isoos) [inline app/lib/account/session_cookie.dart]: Nit: We have `parseCookieHeader` in `shared/utils.dart`.

Comment (jonasfj) [reply app/lib/account/session_cookie.dart]: Okay, did a few tweaks and adopted this..

**Pull request merged**